pennylane/transforms/compile.py

```diff
@@ -167,10 +168,24 @@ def qfunc(x, y, z):
 
     """
 
-    for p in pipeline:
-        p_func = p.func if isinstance(p, partial) else p
-        if not isinstance(p_func, TransformDispatcher):
-            raise ValueError("Invalid transform function {p} passed to compile.")
+    # Convert pipeline to CompilePipeline if it's a tuple or list
+    if isinstance(pipeline, (tuple, list)):
+        # Legacy format: tuple/list of TransformDispatcher or partial(TransformDispatcher)
+        validated_transforms = []
+        for p in pipeline:
+            p_func = p.func if isinstance(p, partial) else p
+            if not isinstance(p_func, TransformDispatcher):
+                raise ValueError(f"Invalid transform function {p} passed to compile.")
+            validated_transforms.append(p)
+        # Convert to partial calls that we apply one by one later
+        # Keep as list for now, will be applied in the transform loop below
+    elif isinstance(pipeline, CompilePipeline):
+        # New format: CompilePipeline object - validation happens inside CompilePipeline
+        pass
+    else:
```

Comment:
I think it would be a bit clearer to replace these two conditional branches with one, since the first one is just empty. This is less reading and expresses what you are looking for clearly.
```suggestion
    elif not isinstance(pipeline, CompilePipeline):
```